### PR TITLE
Order Creation: Align text fields in the address form

### LINKED ISSUE #6018: Order Creation: Spaces initially do not appear when typing in Customer Details
**Describe the bug**
When adding or editing Customer Details, spaces do not appear until another character is typed after the space. @ealeksandrov I know this area is still WIP, not sure if you're aware of this issue.

**To Reproduce**
1. From Orders, tap "+" and choose "Create Order".
2. Tap "Add customer"
3. Type in some customer details. As you do, add some spaces. This is most noticeable when typing an address.
4. Note that, when you type a space, you do not see the space appear. When type the next character, you'll see the space and the new character.

**Video**

https://user-images.githubusercontent.com/15107387/151795303-f6a9446e-a1f1-4935-a293-6f4a64801f45.MP4

**Expected behavior**
I expected to see spaces immediately when typing in customer details.

**Isolating the problem (mark completed items with an [x]):**
- [ ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**Mobile Environment**

- Device: iPhone 12 mini
- iOS version: 14.8.1
- WooCommerce iOS version: 8.4 (33012)

**Tested on Pressable w/Jetpack and on a WordPress.com Atomic site.**

### PR DESCRIPTION
**Pull request by ealeksandrov**
Closes: https://github.com/woocommerce/woocommerce-ios/issues/6018.

## Description

This PR updates Address Form to align textfields vertically. This also fixes issue with missing trailing spaces input.

## Changes
- Adds `MaxWidthPreferenceKey` to a title view in `TitleAndTextFieldRow`.
- Updates `SingleAddressForm` to store shared (max) title width and inject it back into all fields.

### Known limitations
- `TitleAndValueRow` (Country and City fields) wasn't updated yet, but it doesn't have "missing spaces" bug and can be updated in a same way after testing and approval of this PR.

## Testing

1. Open Address Form from New Order or Order Details flows.
2. Enter some data in different fields and confirm that alignment is correct.
3. Update dynamic type size and recheck the layout.

## Screenshots

before|after
--|--
Simulator Screen Shot|Simulator Screen Shot
Simulator Screen Shot|Simulator Screen Shot

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/TitleAndTextFieldRow.swift [hunk 1]
```diff
@@ -58,6 +64,33 @@ private extension TitleAndTextFieldRow {
...
+        }
+    }
+}
+
+private struct MaxWidthModifier: ViewModifier {
+    private var sizeView: some View {
+        GeometryReader { geometry in
+            Color.clear
+                .preference(key: MaxWidthPreferenceKey.self,
+                            value: geometry.size.width)
```
Comment (ealeksandrov) [inline WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/TitleAndTextFieldRow.swift]: Not a fan of helper views, but wasn't able to find a better solution.
There are "Anchor preferences" that implicitly read view geometry, but they are more complicated and doesn't provide single "width" or "layout guide" value. They fit best to convert rect/point from one view context to another, but we don't need it here.

**Review by Ecarrion — APPROVED** [review]
Smart solution @ealeksandrov! It feels a little bit magical and somehow disconnected because `MaxWidthPreferenceKey` is read in one view and read from another in a bottom-up style but I don't have any better suggestion 😅 

I wonder if something similar could have been achieved by environment values? 🤔 

Aside from ☝️, I noticed a couple of small UI Issues.

1. If we use **Dynamic Type XXL** the labels overlap a bit

Screen Shot 2022-04-13 at 10 16 45 AM

2. It looks a bit odd w
Screen Shot 2022-04-13 at 10 17 38 AM
hen the country has no states to select.

**Comment by ealeksandrov** [comment]
Thanks for the review @Ecarrion!

Agree about magical feeling 🙂 From what I understand - environment values are about passing data from top to bottom layers. Here we need to report data from bottom views to the parent.

1. The reason for this issue was just a wrong order of 2 lines 😄
Dynamic type sizing was already handled, it was a mistake preventing live update - dynamic width getter was applied *after* static width setter. Fixed in f37d7b4.

regular size|Dynamic Type XXL
--|--
Simulator Screen Shot - 1|Simulator Screen Shot - 2

2. I plan to update layout for `TitleAndValueRow` separately, after we're happy with current implementation. It should be straightforward to add the same preference, modifier and binding.

**Comment by Ecarrion** [comment]
> Agree about magical feeling

Maybe it will feel less magical if we create a private custom row inside the form where it's clearly visible (because it is in the same file) where the preference width key is being set.

What do you think?

**Comment by ealeksandrov** [comment]
@Ecarrion but that will prevent reuse of this preference in other screens.
Also we need same update to `TitleAndValueRow` and duplicating second field will clutter single file a lot.

I've applied same change for `TitleAndValueRow`, moved preference into a separate file and added some comments ([commits subset](https://github.com/woocommerce/woocommerce-ios/pull/6625/files/000016528d6631d0eba41e3c5d56b5cac31eb330..889967ef91ec20e0c290370a498ad6fd677cb772)).
Do you think it's better or we can add more structure and/or clarifications?

regular size|Dynamic Type XXL
--|--
Simulator Screen Shot - 1|Simulator Screen Shot - 2

**Review by Ecarrion — APPROVED** [review]

**Pull request merged**